url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/drafts/Draft_10.java

```diff
@@ -0,0 +1,397 @@
+import minewebsocket.org.java_websocket.exceptions.NotSendableException;
+import minewebsocket.org.java_websocket.framing.CloseFrameBuilder;
+import minewebsocket.org.java_websocket.framing.FrameBuilder;
+import minewebsocket.org.java_websocket.framing.Framedata;
+import minewebsocket.org.java_websocket.framing.Framedata.Opcode;
+import minewebsocket.org.java_websocket.framing.FramedataImpl1;
+import minewebsocket.org.java_websocket.handshake.ClientHandshake;
+import minewebsocket.org.java_websocket.handshake.ClientHandshakeBuilder;
+import minewebsocket.org.java_websocket.handshake.HandshakeBuilder;
+import minewebsocket.org.java_websocket.handshake.Handshakedata;
+import minewebsocket.org.java_websocket.handshake.ServerHandshake;
+import minewebsocket.org.java_websocket.handshake.ServerHandshakeBuilder;
+import minewebsocket.org.java_websocket.util.Base64;
+import minewebsocket.org.java_websocket.util.Charsetfunctions;
+
+public class Draft_10 extends Draft {
+
+	private class IncompleteException extends Throwable {
+
+		/**
+		 * It's Serializable.
+		 */
+		private static final long serialVersionUID = 7330519489840500997L;
+
+		private int preferedsize;
+		public IncompleteException( int preferedsize ) {
+			this.preferedsize = preferedsize;
+		}
+		public int getPreferedSize() {
+			return preferedsize;
+		}
+	}
+
+	public static int readVersion( Handshakedata handshakedata ) {
+		String vers = handshakedata.getFieldValue( "Sec-WebSocket-Version" );
+		if( vers.length() > 0 ) {
+			int v;
+			try {
+				v = new Integer( vers.trim() );
+				return v;
+			} catch ( NumberFormatException e ) {
+				return -1;
+			}
+		}
+		return -1;
+	}
+
+	private ByteBuffer incompleteframe;
+	private Framedata fragmentedframe = null;
+
+	private final Random reuseableRandom = new Random();
+
+	@Override
+	public HandshakeState acceptHandshakeAsClient( ClientHandshake request, ServerHandshake response ) throws InvalidHandshakeException {
+		if( !request.hasFieldValue( "Sec-WebSocket-Key" ) || !response.hasFieldValue( "Sec-WebSocket-Accept" ) )
+			return HandshakeState.NOT_MATCHED;
+
+		String seckey_answere = response.getFieldValue( "Sec-WebSocket-Accept" );
+		String seckey_challenge = request.getFieldValue( "Sec-WebSocket-Key" );
+		seckey_challenge = generateFinalKey( seckey_challenge );
+
+		if( seckey_challenge.equals( seckey_answere ) )
+			return HandshakeState.MATCHED;
+		return HandshakeState.NOT_MATCHED;
+	}
+
+	@Override
+	public HandshakeState acceptHandshakeAsServer( ClientHandshake handshakedata ) throws InvalidHandshakeException {
+		// Sec-WebSocket-Origin is only required for browser clients
+		int v = readVersion( handshakedata );
+		if( v == 7 || v == 8 )// g
+			return basicAccept( handshakedata ) ? HandshakeState.MATCHED : HandshakeState.NOT_MATCHED;
+		return HandshakeState.NOT_MATCHED;
+	}
+
+	@Override
+	public ByteBuffer createBinaryFrame( Framedata framedata ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 101-113, lines 115-123 into separate methods.